lisa/tests/staging/scheduler/util_tracking.py

```diff
@@ -0,0 +1,377 @@
+
+        return cls(res_dir, plat_info)
+
+    @LoadTrackingAnalysis.df_tasks_signals.used_events
+    @requires_events('sched_switch', 'sched_util_est_cfs', 'sched_util_est_task')
+    def get_task_sched_signals(self, trace, cpu, task_name):
+        """
+        Get a :class:`pandas.DataFrame` with the sched signals for the workload task
+
+        This examines scheduler utilization tracking trace events. You will need a
+        target kernel that includes the required events.
+
+        :returns: :class:`pandas.DataFrame` with a column for each signal for
+          the workload task
+        """
+        df = trace.analysis.load_tracking.df_tasks_signals()
+        df = df[df['comm'] == task_name]
+        window = self.get_task_window(trace, task_name, cpu)
+        df = select_window(df, window)
+
+        # Normalize the signal with the detected task execution start
+        df.index -= window[0]
+
+        return df
+
+    @classmethod
+    def check_from_target(cls, target):
+        super().check_from_target(target)
+
+        try:
+            target.plat_info["cpu-capacities"]
+        except KeyError as e:
+            raise CannotCreateError(str(e))
+
+    @property
+    def rtapp_profile(self):
+        return self.get_rtapp_profile(self.plat_info)
+
+class Convergence(UtilTrackingBase):
+    """
+    Basic checks for estimated utilization signals
+
+    **Expected Behaviour:**
+
+    The estimated utilization of a task is properly computed starting form its
+    `util_avg` value at the end of each activation.
+
+    Two signals composes the estimated utlization of a task:
+
+    * `util_est_enqueued` : is expected to match the max between `util_avg` and
+      `util_est_ewma` at the end of the previous activation
+
+    * `util_est_ewma` : is expected to track an Exponential Weighted Moving
+      Average of the `util_avg` signal sampled at the end of each activation.
+
+    Based on these two invariant, this class provides a set of tests to verify
+    these conditions using different methods and sampling points.
+    """
+    task_prefix = 'invariance'
+    @classmethod
+    def get_rtapp_profile(cls, plat_info):
+        big_cpu = list(reversed(plat_info["capacity-classes"]))[0][0]
+
+        profile = {}
+        profile['test_task'] = RTATask()
+
+        # Big task: 15 activations at full utilization
+        profile['test_task'] += Periodic(
+            duty_cycle_pct=75, duration_s=5, period_ms=200, cpus=[big_cpu])
+
+        # Ramp Down
+        profile['test_task'] += Ramp(
+            start_pct=50,
+            end_pct=5,
+            delta_pct=20,
+            time_s=1,
+            period_ms=200,
+            cpus=[big_cpu])
+
+        # Ramp Up
+        profile['test_task'] += Ramp(
+            start_pct=10,
+            end_pct=60,
+            delta_pct=20,
+            time_s=1,
+            period_ms=200,
+            cpus=[big_cpu])
+
+        return profile
```

Comment:
phases can be added together and the profile built at the end:
```
task = (
       Periodic(duty_cycle_pct=75, duration_s=5, period_ms=200, cpus=[big_cpu])
     + Periodic(
            duty_cycle_pct=75, duration_s=5, period_ms=200, cpus=[big_cpu])

        # Ramp Down
        Ramp(
            start_pct=50,
            end_pct=5,
            delta_pct=20,
            time_s=1,
            period_ms=200,
            cpus=[big_cpu])
       +  ....
)
return {'test_task': task}
```

Reply:
Sure... but really, that's mostly taste... ;)